### PR TITLE
[feat] Auto MAC addr detection

### PR DESCRIPTION
**Pull request by baylanger**

### CODE REVIEW

**Comment by raphmur** [comment]
See https://github.com/tesla-local-control/tesla-local-control-addon/pull/32

Anyways, referencing this issue that will be solved with this: https://github.com/tesla-local-control/tesla-local-control-addon/issues/31

File: subroutines.sh [hunk 1]
```diff
@@ -85,9 +99,26 @@ bluetoothctl_read() {
   # Read BLE data from bluetoothctl or an input file
   if [ -z $BLECTL_FILE_INPUT ]; then
     log_debug "Launching bluetoothctl to check for BLE presence"
-    BLECTL_TIMEOUT=11
     set +e
-    BLTCTL_OUT=$(bluetoothctl --timeout $BLECTL_TIMEOUT scan on 2>&1 | grep -v DEL)
+    BLTCTL_OUT=$({
+      if [ $BLTCTL_COMMAND_DEVICES == "true" ]; then
+        bltctlCommands="power on,devices,scan on"
```
Comment (raphmur) [inline subroutines.sh:105]: power on will not cause issues if the bluetooth device is used by something else?

File: subroutines.sh [hunk 2]
```diff
@@ -35,9 +35,23 @@ replace_value_at_position() {
 # Function
 check_presence() {
   TYPE="$1" # BLE MAC, LN
```
Comment (raphmur) [inline subroutines.sh:37]: I don't really see where TYPE is necessary other than logging?

File: subroutines.sh [hunk 3]
```diff
@@ -132,8 +164,18 @@ listen_to_ble() {
...
-      log_debug "$(echo "$BLTCTL_OUT" | grep -E "($BLE_MAC|$BLE_LN)")"
-      check_presence "BLE MAC & LN" "($BLE_MAC|$BLE_LN)"
+      log_debug "BLTCTL_OUT:$(echo "$BLTCTL_OUT" | grep -E "($BLE_MAC|$BLE_LN)")"
+      macAddr=$(check_presence "BLE LN & MAC" $BLE_LN $BLE_MAC)
+
+      # Add the MAC address to BLE_MAC_LIST if it's not already present at the position
+      if [ "$macAddr" != "$BLE_MAC" ]; then
+        eval "BLE_MAC_LIST=\$(echo \$BLE_MAC_LIST | awk '{\$${position}=\"$macAddr\"; print}')"
+        log_debug "listen_to_ble; BLE_MAC_LIST:$BLE_MAC_LIST"
+        [ ! -f $KEYS_DIR/${vin}_macaddr ] && echo $macAddr >$KEYS_DIR/${vin}_macaddr
```
Comment (raphmur) [inline subroutines.sh:174]: good
Then do you check the presence of this file somewhere? Did not find it.

File: env.sh
Comment (raphmur) [inline env.sh:1]: Leaving the BLE_MAC_LIST here will fail?

File: run.sh [hunk 4]
```diff
@@ -10,16 +10,28 @@ echo "[$(date +%H:%M:%S)] Starting... loading /app/env.sh"
...
 
 vin_count=0
 for vin in $VIN_LIST; do
   # Populate BLE Local Names list
   vin_count=$((vin_count + 1))
   BLE_LN=$(vinToBLEln $vin)
-  log_debug "Adding $BLE_LN to BLE_LN_LIST, count $vin_count"
+  log_debug "Adding $BLE_LN to BLE_LN_LIST, vin_count:$vin_count"
   BLE_LN_LIST="$BLE_LN_LIST $BLE_LN"
+  if [ -f $KEYS_DIR/${vin}_macaddr ]; then
```
Comment (raphmur) [inline run.sh:22]: ok found it... Ctrl-F and git search was no bueno

File: subroutines.sh [hunk 5]
```diff
@@ -119,6 +150,7 @@ listen_to_ble() {
   while :; do
     bluetoothctl_read
 
+    BLTCTL_COMMAND_DEVICES=false
     for position in $(seq $n_vins); do
```
Comment (raphmur) [inline subroutines.sh:154]: hmmm I suspect this will raise shellcheck issues?

**Review by raphmur — COMMENTED** [review]
Can you explain the logic? A bit complex for me...
If I understand well:
- you use the LNAME brought by bluetoothctl to check the presence?
- I saw you store the MAC in a file, but dont read it. Normal?
Maybe you are not done. This case, sorry :)

File: subroutines.sh (hunk 1 above)
Comment (baylanger) [reply subroutines.sh:105]: Well... I had to do that for the `scan on` to work on mine. As you can see below, I kept the output cuz it was interesting... I sent the command `devices` and nothing showed up. Did `power on` and then another `devices` command.

```
[bluetooth]# hci0 new_settings: powered bondable ssp br/edr le secure-conn wide-band-speech
[bluetooth]# [CHG] Controller 04:42:1A:5B:2B:00 Pairable: yes
[bluetooth]# AdvertisementMonitor path registered
[bluetooth]# devices
[bluetooth]# power on
[bluetooth]# Changing power on succeeded
[bluetooth]# devices
hci0 type 6 discovering on
[CHG] Controller 04:42:1A:5B:2B:00 Discovering: yes
[NEW] Device 71:D4:D8:AB:FC:A2 71-D4-D8-AB-FC-A2
[NEW] Device 70:72:8E:07:4C:4E 70-72-8E-07-4C-4E
[NEW] Device 61:6A:54:6A:87:89 61-6A-54-6A-87-89
[NEW] Device 5B:62:F4:F9:03:BA 5B-62-F4-F9-03-BA
[NEW] Device 43:8A:B2:DC:87:DE 43-8A-B2-DC-87-DE
[NEW] Device 61:11:DD:7E:0E:45 61-11-DD-7E-0E-45
[NEW] Device 1C:B3:C9:11:77:6C 1C-B3-C9-11-77-6C
[NEW] Device 5D:93:2D:44:4F:D9 5D-93-2D-44-4F-D9
```

If the adapter is powered off, the power on is required. If a light bulb is already on, it won't blow up to "turn it on" again. If the adapter is already powered on, I doubt sending a "power on" has any impact.

File: subroutines.sh (hunk 2 above)
Comment (baylanger) [reply subroutines.sh:37]: True - that is indeed useless today. Will remove it, thank you - again you have good eyes and definitely going in the code :)

The history... in the first iteration of check_presence ( ~ 2 weeks ago) the function was receiving the Type and a pattern Match. The pattern could a MAC, BLE LN or both. Thus the reason to pass the type.

By adding in this PR a default Mac address `FF:FF....:FF` , there's a defined Mac address (valid or not) for each VIN/BLE LN that can be pass to the function. So... Match=$2 is now replaced by BLE LN=$2  * MAC addr=$3.

File: subroutines.sh (hunk 3 above)
Comment (baylanger) [reply subroutines.sh:174]: Yes in run.sh `for loop` - it adds either the real MAC address or "fake" ``FF:FF...FF`.

File: subroutines.sh (hunk 5 above)
Comment (baylanger) [reply subroutines.sh:154]: It is defined in env.sh and we load that first.

**Review by iainbullock — COMMENTED** [review]
Excellent work and a very useful new feature, many thanks
A lot of changes. Have either of you tested it before we merge into #main? 
I could create a _docker branch and use it to test, however I might not complete it today

**Comment by baylanger** [comment]
> Excellent work and a very useful new feature, many thanks A lot of changes. Have either of you tested it before we merge into #main? I could create a _docker branch and use it to test, however I might not complete it today

Yes I have been testing throughout the dev process but more tester is always good. I did create the addon PR but not for Docker. Will do if you don't have time and that I get the time.

**Comment by iainbullock** [comment]
Ok shall we merge into core? I don't think there's a lot to do for _docker from what I can see. Eg
Remove BLE_MAC_LIST references
Update installation docs
Update CHANGELOG

Anything else? I can do that later no problem

File: env.sh [hunk 6]
```diff
@@ -34,14 +34,14 @@ export PRESENCE_DETECTION_LOOP_DELAY=${PRESENCE_DETECTION_LOOP_DELAY:-120}
 export PRESENCE_DETECTION_TTL=${PRESENCE_DETECTION_TTL:-240}
 
 export BLE_LN_REGEX='S[0-9A-Fa-f]{16}C'
+export BLTCTL_COMMAND_DEVICES=false
 export KEYS_DIR=/share/tesla_ble_mqtt
-export MAC_REGEX='([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})'
+export MAC_REGEX='([0-9A-Fa-f]{2}[:]){5}([0-9A-Fa-f]{2})'
```
Comment (raphmur) [inline env.sh:39]: Do we still need the MAC regex?

**Review by raphmur — COMMENTED** [review]
Me likey likey!
Happy to test.

**Review by raphmur — APPROVED** [review]

**Review by iainbullock — APPROVED** [review]

**Pull request merged**